core/include/cubos/core/net/tcp_listener.hpp

```diff
@@ -0,0 +1,72 @@
+/// @file
+/// @brief Class @ref cubos::core::net::TcpListener.
+/// @ingroup core-net
+
+#pragma once
+
+#include <cubos/core/api.hpp>
+#include <cubos/core/net/address.hpp>
+#include <cubos/core/net/tcp_stream.hpp>
+
+#ifdef _WIN32
+#include <winsock2.h>
+#pragma comment(lib, "Ws2_32.lib")
+#else
+#include <sys/socket.h>
+// Socket type. Differs from Windows (SOCKET) and POSIX (int).
+#define SOCKET int
+// Represents an invalid socket.
+#define INVALID_SOCKET (-1)
+#endif
+
+namespace cubos::core::net
+{
+    class CUBOS_CORE_API TcpListener final
+    {
+    public:
+        /// @brief Create an empty TCP listener, ready for connecting.
+        TcpListener();
+
+        /// @brief Deconstructs by calling @ref close.
+        ~TcpListener();
+
+        /// @brief Forbid copy construction.
+        TcpListener(const TcpListener&) = delete;
+
+        /// @brief Forbid assignment.
+        TcpListener& operator=(const TcpListener&) = delete;
+
+        /// @brief Start listening for connection at specified `address` and `port`.
+        /// @param address Destination address.
+        /// @param port Destination port.
+        /// @param maxQueueLength Maximum connection requests.
+        /// @return Whether start listening was successfull.
+        bool listen(const Address& address, uint16_t port, int maxQueueLength = SOMAXCONN);
+
+        /// @brief Accepts a new connection.
+        /// @param[out] stream Buffer to store the connection stream.
+        /// @return Whether accepting the connection and creating the stream was successfull.
+        /// @note This function will block until a connection is made.
+        bool accept(TcpStream& stream) const;
+
+        /// @brief Close the listener socket.
+        void close();
+
+        /// @brief Check if connection is valid.
+        /// @return Whether inner sock is connected.
+        inline bool connected() const
+        {
+            return mSock != INVALID_SOCKET;
+        }
```

Comment:
Unsure `connected` is the right work here. Maybe `valid`? As there isn't really a connection.